web/pandas/pdeps/0014-string-dtype.md

```diff
@@ -0,0 +1,385 @@
+`pd.NA` (but this PDEP proposes a better naming scheme, see the "Naming"
+subsection below).
+
+This last dtype variant is what you currently (pandas 2.2) get for string data
+when enabling the ``future.infer_string`` option (to enable the behaviour which
+is intended to become the default in pandas 3.0).
+
+## Proposal
+
+To be able to move forward with a string data type in pandas 3.0, this PDEP proposes:
+
+1. For pandas 3.0, we enable a "string" dtype by default, which will use PyArrow
+   if installed, and otherwise falls back to an in-house functionally-equivalent
+   (but slower) version.
+2. This default "string" dtype will follow the same behaviour for missing values
+   as our other default data types, and use `NaN` as the missing value sentinel.
+3. The version that is not backed by PyArrow can reuse (with minor code
+   additions) the existing numpy object-dtype backed StringArray for its
+   implementation.
+4. We update installation guidelines to clearly encourage users to install
+   pyarrow for the default user experience.
+
+Those string dtypes enabled by default will then no longer be considered as
+experimental.
+
+### Default inference of a string dtype
+
+By default, pandas will infer this new string dtype instead of object dtype for
+string data (when creating pandas objects, such as in constructors or IO
+functions).
+
+The existing `future.infer_string` option can be used to opt-in to the future
+default behaviour:
+
+```python
+>>> pd.options.future.infer_string = True
+>>> pd.Series(["a", "b", None])
+0      a
+1      b
+2    NaN
+dtype: string
+```
+
+This option will be expanded to also work when PyArrow is not installed.
+
+### Missing value semantics
+
+As mentioned in the background section, the original `StringDtype` has used
+the experimental `pd.NA` sentinel for missing values. In addition to using
+`pd.NA` as the scalar for a missing value, this essentially means
+that:
+
+- String columns follow ["NA-semantics"](https://pandas.pydata.org/docs/user_guide/missing_data.html#na-semantics)
+  for missing values, where `NA` propagates in boolean operations such as
+  comparisons or predicates.
+- Operations on the string column that give a numeric or boolean result use the
+  nullable Integer/Float/Boolean data types (e.g. `ser.str.len()` returns the
+  nullable `'Int64"` / `pd.Int64Dtype()` dtype instead of the numpy `int64`
+  dtype (or `float64` in case of missing values)).
+
+However, up to this date, all other default data types still use NaN semantics
+for missing values. Therefore, this proposal says that a new default string
+dtype should also still use the same default missing value semantics and return
+default data types when doing operations on the string column, to be consistent
+with the other default dtypes at this point.
+
+In practice, this means that the default `"string"` dtype will use `NaN` as
+the missing value sentinel, and:
```

Comment:
Interestingly, I don't think the current `object` behavior always uses `NaN`:
```python
>>> sn = pd.Series(["abc", "defg", "hijkl"])
>>> sn
0      abc
1     defg
2    hijkl
dtype: object
>>> sn.shift(1)
0    None
1     abc
2    defg
dtype: object
>>> sn.shift(1).iloc[0] is None
```
So at least with the `shift()` operation, the "missing value" is `None`, not `np.nan`

Reply:
Yeah, we are not really consistent with this. When created, we will consider any null-like (None, NaN, NaT, NA) as a missing value in an object-dtype column, but for methods introducing them I would have hoped we were consistent. But apparently not:

```python
>>> ser = pd.Series(["a", "b", "c"], dtype=object)
>>> ser.shift(1)
0    None
1       a
2       b
dtype: object
>>> ser.reindex([1, 2, 3])
1      b
2      c
3    NaN
dtype: object
```

And to be clear, this will all be NaN with the future string dtype (whether converted to NaN upon construction, or ensuring we use NaN for missing values introduced in methods)

Reply:
The fact that we're not consistent today does suggest that with this proposal (using `np.nan` as the default missing value indicator for strings), that there _will_ be a behavior change for people using `dtype=object` today with strings, because we'd replace `None` with `np.nan` in `shift()` (and maybe elsewhere)

Reply:
Mentioned this in the backwards compatibility section